README.txt

======
README
======

*$Id$*


OOo2DBK is an OpenOffice.org document to DocBook XML converter. It can produce
both DocBook article and DocBook book documents.

The online reference presenting all the available DocBook elements can is
http://docbook.org/tdg/en/html/

OOo2DBK can at the moment only process files generated by the OpenOffice.org
1.1.x series. Future versions of OOo2DBK will process files generated by the
OpenOffice.org 2.x series.


Prerequisites
=============

To run ooo2dbk on your system you need on your system:

  * Python 2.3 (because of the dependence to the optparse module)

  * any supported XSLT processor (check in config.xml)
    ooo2dbk defaults to using the faster xsltproc program
    http://xmlsoft.org/XSLT/xsltproc2.html But you can use any other XSLT
    processor, such as Saxon.


Examples
========

1. To produce a DocBook XML article out of an OpenOffice.org document:
$ ooo2dbk --article --zipfile=document.zip document.sxw

2. To produce a DocBook XML book out of an OpenOffice.org document:
$ ooo2dbk --book --zipfile=document.zip document.sxw


Styles
======

The following styles can be used in OpenOffice.org and map to the corresponding
DocBook tags:

Abstract		abstract
Acronym			acronym
Attribution		attribution
Author			author
Authorblurb		authorblurb
Bibliocoverage		bibliocoverage
Bibliomisc		bibliomisc
Bibliorelation		bibliorelation
Bibliosource		bibliosource
Blockquote		blockquote
Citation		citation
Command			command
Computeroutput		computeroutput
Copyright		copyright
Corpauthor		corpauthor
Definition		element used for variablelist
Editor			editor
Email			email
Emphasis		emphasis
Epigraph		epigraph
Example			literal
Filename		filename
Firstname		firstname
Frame			element used for figure
Graphics		element used for figure
Guillemets		element used to mark quotes we don't want to treat
Heading			title
Holder			holder
Honorific		honorific
Jobtitle		jobtitle
Keyword			keyword
Keywordset		keywordset
Legalnotice		legalnotice
Lineage			lineage
Lineannotation		lineannotation
Literal			literal
Literallayout		literallayout
Note			note
Option			option
Orgname			orgname
Othercredit		othercredit
Othername		othername
Preformatted Text	programlisting
ProgramListing		programlisting
Pubdate			pubdate
Pubsnumber		pubsnumber
Quotations		blockquote
Remark			remark
Replaceable		replaceable
Source Text		literal
Strong Emphasis		emphasis role="strong"
Subtitle		subtitle
Surname			surname
Teletype		literal
Term			term
Title			title
Userentry		userinput
User Entry		userinput
Year			year


Problems and debugging
======================

In case of problems run ooo2dbk with the "--verbose" option.


Is it possible to run ooo2dbk on Microsoft Windows ?
====================================================

If you get the following problem:

C:\WINDOWS\Bureau\ooo2dbk-1.0.2-1>c:\python23\python ooo2dbk --book --zipfile=document.zip document.sxw

Commande ou nom de fichier incorrect
Traceback (most recent call last):
  File "ooo2dbk", line 587, in ?
    exec_args()
  File "ooo2dbk", line 131, in exec_args
    deltemp=options.flatxml, xslParams=xslParams)
  File "ooo2dbk", line 560, in convert
    make_zip()
  File "ooo2dbk", line 508, in make_zip
    zarchive.write(sDocbookXML, pjoin(archive_top_dir, docbook_fname))
  File "C:\PYTHON23\lib\zipfile.py", line 391, in write
    st = os.stat(filename)
OSError: [Errno 2] No such file or directory: 'C:\\WINDOWS\\Bureau\\ooo2dbk-1.0.
2-1\\document.docb.xml'

The problem is that there is no 'xsltproc' executable installed.
To solve the problem you need to install it.

1) Download:
  - ftp://ftp.zlatkovic.com/pub/libxml/libxml2-2.6.15.win32.zip
  - ftp://ftp.zlatkovic.com/pub/libxml/libxslt-1.1.12.win32.zip
  - ftp://ftp.zlatkovic.com/pub/libxml/zlib-1.2.1.win32.zip
  - ftp://ftp.zlatkovic.com/pub/libxml/iconv-1.9.1.win32.zip

2) Create directory somewhere on filesystem, for example, c:\libxml

3) Put c:\libxml into system PATH on windows
(On WinXP you can achieve this with following: Start -> Control Panel -> System,
then choose tab "Advanced" and press "Environment variables".  You will see
"System variables", find there "Path" and append "c:\libxml" at the end)

4) From first 3 above downloaded zip files copy _all_ contents of "bin"
directory into "c:\libxml". As for iconv-1.9.1.win32.zip - copy iconv.dll from
"lib" and "iconv.exe" from "util" into "c:\libxml".

5) Restart shell from which you ran ooo2dbk (this will take into account
modified PATH) and try again.

This has been tested with both Microsoft Windows XP (SP2) and Windows 2000.

Creating standalone executable on Microsoft Windows
===================================================

You can build standalone console application with the help of 'py2exe' -
http://py2exe.sourceforge.net

It's a version 0.6.3 at the moment of the writing.

After installing py2exe execute in ooo2dbk source directory:

python setup.py py2exe -b 1

As a result you'll get 'dist' directory containing:
library.zip, ooo2dbk.exe, w9xpopen.exe that should be distributed together
to run oo2dbk.exe

Don't forget that you'll also need at least two additional files from source
distribution: ooo2dbk.xml and ooo2dbk.xsl for proper use of ooo2dbk.exe

To make ooo2dbk.exe available system wide, put path to directory containing
that executable into system PATH variable.
Good luck!

ooo2dbk along with a OpenOffice.org instance
============================================

If you want to use the OLE objects to image conversion feature you need to
launch the ooo2dbk script with the python shipped with OpenOffice.org.
Since OpenOffice.org 1.1.x series ship with Python 2.2 you need to take extra
steps to make this combination possible.

The solution is to copy the Python 2.3 specific modules that are needed by
ooo2dbk to the OpenOffice.org shipped Python module location:
/usr/local/openoffice-1.1.4/program/python-core/lib$ cp /usr/lib/python2.3/optparse.py .
/usr/local/openoffice-1.1.4/program/python-core/lib$ cp /usr/lib/python2.3/textwrap.py .



CHANGES
Requires
~~~~~~~~
-
New features
~~~~~~~~~~~~
-
Bug fixes
~~~~~~~~~
- Fixed reference in the comments and messages to config.xml file which is now a
  constant that may change.
New internal features
~~~~~~~~~~~~~~~~~~~~~
- 


test_ooo2dbk.py
#!/usr/bin/python
# (C) Copyright 2003-2006 Nuxeo SAS <http://nuxeo.com>
#
# Authors:
# Laurent Godard (lgodard@indesko.com)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# See ``COPYING`` for more information
#
# $Id$

import zipfile
import os, os.path
import tempfile

test_cases= [{'input_file': 'tests/input/test.sxw', 
              'expected_dir': 'tests/output/sxw.out'},

             {'input_file': 'tests/input/test.odt',
             'expected_dir': 'tests/output/odt.out'},
            ]

def ooo2dbk_tester(input_file, expected_dir):

    target_dir = tempfile.mkdtemp(prefix = 'ooo2dbk-test-')

    command = './ooo2dbk -d %s %s' % (os.path.join(target_dir,'test.docb.xml'),
                                                   input_file)
    print command
    os.system(command)

    status = True

    # are all the images processed ?
    expected_images_dir = os.path.join(expected_dir, 'images')
    for image in os.listdir(expected_images_dir):
        image_to_find = os.path.join(expected_images_dir, image)
        if os.path.isfile(image_to_find): #to avoid .svn dir
            expect = os.path.join(target_dir, 'images',image)
            if not os.path.isfile(expect):
                status = status and False
                print 'image %s not found' % image

    # compare docbook files ?
    f = open(os.path.join(expected_dir, 'test.docb.xml'))
    model = '\n'.join(f.readlines())
    f.close()
    g = open(os.path.join(target_dir, 'test.docb.xml'))
    computed = '\n'.join(g.readlines())
    g.close()

    if computed != model:
        status = False
        print "docbook files are different !"
        command = "diff %s %s" % (os.path.join(expected_dir, 'test.docb.xml'),
                                  os.path.join(target_dir, 'test.docb.xml'))
        os.system(command)

    # clean temp file
    os.system('rmdir --ignore-fail-on-non-empty %s' % target_dir)

    return status

if __name__ == '__main__':

    for test in test_cases:
        print '--------------------'
        status = ooo2dbk_tester(test['input_file'], test['expected_dir'])
        if status:
            result = 'PASS'
        else:
            result = 'FAIL'
        print "\n%s on %s" %(result, test['input_file'])

__init__.py
#==================================================================
# OOo2DBK : OpenOffice.org Writer document to DocBook XML converter
#==================================================================
# (C) Copyright 2006 Nuxeo SAS <http://nuxeo.com>
# (C) Copyright Eric Bellot <ebellot@netcourrier.com>
#
# Authors:
# Ruslan Spivak (Nuxeo)
# M.-A. Darche (Nuxeo)
# Eric Bellot <ebellot@netcourrier.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# See ``COPYING`` for more information
#
# $Id$
#from ooo2dbk import execArgs, convert



ole2img.py
#!/usr/bin/python
# Copyright (c) 2004-2005 Nuxeo SARL <http://nuxeo.com>
# Authors:
# Laurent Godard <lgodard@indesko.com>
# M.-A. Darche (Nuxeo)
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 2 as published
# by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
# 02111-1307, USA.
#
# See ``COPYING`` for more information
#
# $Id$

# UNO
import uno, unohelper
from com.sun.star.beans import PropertyValue
from com.sun.star.connection import NoConnectException

import sys
import getopt
import os.path
from os import sep
from os import remove
import time

VERSION = '0.6'

#filter parameters
FILTER_PARAMS = {
    'png': ('.png', 'image/png'),
    'svg': ('.svg', 'image/svg+xml'),
    'jpg': ('.jpg', 'image/jpeg'),
    'tiff': ('.tiff', 'image/tiff'),
    'eps': ('.eps', 'application/postscript'),
    'bmp-ms': ('.bmp', 'image/x-MS-bmp'),
    'bmp-portable': ('.bmp', 'image/x-portable-bitmap'),
    'gif': ('.gif', 'image/gif'),
    }


def usage():
    lst_formats = str(FILTER_PARAMS.keys())
    print "usage: ole2img [options] openoffice.org-file"
    print
    print "options:"
    print "  --target    target directory (mandatory)"
    print "  --oooserverhost the name of server running OpenOffice.org (defaults to localhost)"
    print "  --oooserverport the port of the server running OpenOffice.org (defaults to 2002)"
    print "  --format    exported image format (defaults to png)"
    print "              %s" % lst_formats.replace("'","")
    print "  --version   show version and exit"
    print "  --help      show this help message and exit"
    print
    print "Example:"
    print "~/OpenOffice.org1.1.2/program/python ole2img --format png --target ~/outputdir/ ~/exampleFile.sxw"
    print
    print "OpenOffice.org must run in listen mode"
    print './soffice "-accept=socket,host=localhost,port=2002;urp;"'


def version():
    print "Converts all OLE objects of a .sxw OpenOffice.org document as images in a target directory"
    print "ole2img version: " + VERSION


def exec_args():
    """Analyze command line arguments.
    """
    source = None
    target = None
    # Default server having running OpenOffice.org instance(s) is localhost
    oooserver_host = 'localhost'
    oooserver_port = '2002'
    # Default format is PNG
    format = 'png'
    try:
        opts, args = getopt.getopt(sys.argv[1:], '',
                                   ['target=',
                                    'oooserverhost=',
                                    'oooserverport=',
                                    'format=',
                                    'version', 'help'])
    except:
        usage()
        # Command line syntax errors return the error code "2"
        sys.exit(2)

    if len(args) == 1:
        source = args[0]

    for opt in opts:
        if opt[0] == '--target':
            target = opt[1]
        elif opt[0] == '--oooserverhost':
            oooserver_host = opt[1]
        elif opt[0] == '--oooserverport':
            oooserver_port = opt[1]
        elif opt[0] == '--format':
            format = opt[1]
        elif opt[0] == '--version':
            version()
            sys.exit(0)
        elif opt[0] == '--help':
            usage()
            sys.exit(0)
    error = False

    if source == None:
        error = True
        print "\nOne and only one OpenOffice.org file must be given"
    else:
        source = os.path.abspath(os.path.expanduser(source))
        if not os.path.isfile(source):
            error = True
            print "\n--source Invalid file"

    if target == None:
        error = True
        print "\n--target Missing Parameter"
    else:
        target = os.path.abspath(os.path.expanduser(target))
        if not os.path.isdir(target):
            error = True
            print "\n--target Invalid directory"

    if not FILTER_PARAMS.has_key(format):
        error = True
        print "\n--format Undefined value"

    if error:
        usage()
        # Command line syntax errors return the error code "2"
        sys.exit(2)

    #add Os separator if missing to target
    if target[len(target) - 1] != os.sep:
        target += os.sep

    return source, target, oooserver_host, oooserver_port, format


def ole2img(source, target,
            oooserver_host='localhost', oooserver_port='2002', format='png'):
    #fileter parameters
    extension, theFilter = FILTER_PARAMS[format]

    #Connect to OOo

    # get the uno component context from the PyUNO runtime
    localContext = uno.getComponentContext()
    # create the UnoUrlResolver
    resolver = localContext.ServiceManager.createInstanceWithContext(
                                'com.sun.star.bridge.UnoUrlResolver', localContext )

    # connect to the running office
    try:
        print "Connecting to server %s:%s ..." % (oooserver_host, oooserver_port)
        ctx = resolver.resolve(
            'uno:socket,host=%s,port=%s;urp;StarOffice.ComponentContext'
            % (oooserver_host, oooserver_port))
    except NoConnectException:
        print "Unable to connect to OpenOffice.org instance"
        sys.exit(1)

    smgr = ctx.ServiceManager

    # get the central desktop object
    desktop = smgr.createInstanceWithContext('com.sun.star.frame.Desktop', ctx)

    # Now connected to OOo

    # load source file
    args = (PropertyValue('Hidden', 0, True, 0),)
    url = unohelper.systemPathToFileUrl(source)
    sourceDoc = desktop.loadComponentFromURL(url, '_blank', 0, args)

    # Hack/Bug : Needed to sanitize the object hierarchy for use with CurrentController
    url = unohelper.systemPathToFileUrl(target + 'temp.sxw')
    sourceDoc.storeToURL(url,())
    os.remove(target + 'temp.sxw')

    oGraphic=smgr.createInstanceWithContext(
                            'com.sun.star.drawing.GraphicExportFilter', ctx)

    #dispatcher for copy-paste
    dispatcher = smgr.createInstanceWithContext(
                    'com.sun.star.frame.DispatchHelper', ctx)

    #creates an hidden draw docuement
    args = (PropertyValue('Hidden', 0, True, 0),)
    url = 'private:factory/sdraw'
    drawDoc = desktop.loadComponentFromURL(url, '_blank', 0, args)

    theOleObjects = sourceDoc.EmbeddedObjects


    for i in range(theOleObjects.Count):
        oleObject = theOleObjects.getByIndex(i)
        print oleObject.Name
        # Selection
        sourceDoc.CurrentController.select(oleObject)

        # XXX : OOo
        # strange random bug on importing documents from previous OOo version
        # sometimes the current object may not be selected
        # the selection remains on the previous in the loop

        # May be not needed any more as done previously ?

        # Verify the selection is ok
        obj = sourceDoc.CurrentController.getSelection()
        if obj.Name != oleObject.Name:
            # Save the file and select again the object
            # Updates the object and the can be selected
            url = unohelper.systemPathToFileUrl(target + 'temp.sxw')
            sourceDoc.storeToURL(url,())
            os.remove(target + 'temp.sxw')
            sourceDoc.CurrentController.select(theOleObjects.getByIndex(i))
            obj = sourceDoc.CurrentController.Selection

        # if the previous did not work
        # creates an image telling a problem occured
        if obj.Name != oleObject.Name:
            print "unable to process " + oleObject.Name
            print 1/0
            drawingShape = drawDoc.createInstance("com.sun.star.drawing.TextShape")
            theSapeSize = drawingShape.Size
            drawDoc.DrawPages.getByIndex(0).add(drawingShape)
            theSapeSize.Width = 5000
            theSapeSize.Height = 2500
            drawingShape.setSize(theSapeSize)
            drawingShape.String = oleObject.Name + " - Unable to process OLE object"
            drawingShape.CharColor = 255
            drawingShape.CharHeight = 8
            objDraw = drawingShape
        else:
        # Normal behaviour
            # Copy
            dispatcher.executeDispatch(sourceDoc.CurrentController.Frame,
                                       '.uno:Copy',
                                       '',
                                       0,
                                       ())
            # Paste into draw
            dispatcher.executeDispatch(drawDoc.CurrentController.Frame,
                                       '.uno:Paste',
                                       '',
                                       0,
                                       ())

            # Rename draw object
            objDraw = drawDoc.CurrentController.Selection

            # Export
            oGraphic.setSourceDocument(objDraw)
            url = unohelper.systemPathToFileUrl(target + oleObject.Name + extension)
            argsExport = (PropertyValue('URL' , 0 , url, 0 ),
                          PropertyValue('MediaType' , 0 , theFilter, 0 ))
            oGraphic.filter(argsExport)

    # Close files
    drawDoc.close(True)
    sourceDoc.close(True)


# Shell access
if __name__ == "__main__":
    startTime = time.time()
    source, target, oooserver_host, oooserver_port, format = exec_args()
    print "Exporting into %s format to target %s..." % (format, target)
    ole2img(source, target, oooserver_host, oooserver_port, format)
    endTime = time.time()
    duration = round(endTime - startTime, 2)
    print 'duration : %s sec' % duration
